manual/io/index.md

```diff
@@ -56,784 +46,433 @@ root [1] TFile f1("my.root");
+The schema evolution C++ API consists of the following classes:
 
-To create an XML file, specify a filename with an `.xml` extension when calling [TFile::Open()](https://root.cern/doc/master/classTFile.html#ad8870806a04da2c2f4aa02bee4ec6833){:target="_blank"}. `TFile::Open()` recognizes that you are trying to open an XML file and returns a {% include ref class="TXMLFile" %} object. When a XML file is open in write mode, you can use [TObject::Write()](https://root.cern/doc/master/classTObject.html#a19782a4717dbfd4857ccd9ffa68aa06d){:target="_blank"} to write an object into the XML file.
+- [TSchemaRuleSet](https://root.cern/doc/master/classROOT_1_1Detail_1_1TSchemaRuleSet.html){:target="_blank"}: objects of this type manage the sets of rules and ensure their consistency. There can be no conflicting rules in the rule sets. The rule sets are owned by the {% include ref class="TClass" %} objects corresponding to the target classes defined in the rules and can be accessed using [`TClass::GetSchemaRules()`](https://root.cern/doc/master/classTClass.html#a8ffa89c0d9362cd77d957fed12ea9fca){:target="_blank"} and [`TClass::AdoptSchemaRules()`](https://root.cern/doc/master/classTClass.html#a484ed223d16ffbf0d35cee567e6f6225){:target="_blank"}.
+- [TSchemaRule](https://root.cern/doc/master/classROOT_1_1TSchemaRule.html){:target="_blank"}: it represent the rules and their fields have exactly the same meaning as the ones of rules specified in the dictionaries (see [above](#specifying-io-customization-rules-in-a-linkdef-file)).
 
-_**Example**_
+#### Schema evolution with custom streamers
+
+If you have written your own `Streamer` as described in [Custom streamers](#custom-streamers), you will have to manually add code for each version and manage the evolution of your class. When you add or remove data members, you must modify the `Streamer` by hand. ROOT assumes that you have increased the class version number in the `ClassDef` statement and introduced the relevant test in the read part of the Streamer. For example, if a new version of the `Event` class above includes a new member: `Int_t fNew` the `ClassDef` statement should be changed to `ClassDef(Event,2)` and the following lines should be added to the read part of the `Streamer`:
 
 {% highlight C++ %}
-// Example of a session saving a histogram to a XML file.
-   TFile *f = TFile::Open("Example.xml","recreate");
-   TH1F *h = new TH1F("h","test",1000,-2,2)
-   h->FillRandom("gaus");
-   h->Write();
-  delete f;
-
-// Example of a session saving a histogram to a XML file.
-   TFile *f = TFile::Open("Example.xml");
-   TH1F *h = (TH1F*)f->Get("h");
-   h->Draw();
-   canvas->Print("Example.xml");
+   if (R__v > 1) R__b >> fNew;
+   else fNew = 0;        // set to some default value
```

Comment:
@pcanal That should be it, no?
```suggestion
   // `buf` below is the TBuffer& argument to the `Streamer` function
   UInt_t R__s, R__c;
   Version_t R__v = buf.ReadVersion(&R__s, &R__c);

   Event::Class()->ReadBuffer(buf, this);
   if (R__v > 1) buf >> fNew;
   else fNew = 0;        // set to some default value

   buf.CheckByteCount(R__s, R__c, Event::IsA());
```

Reply:
humm .. looks odd.

Reply:
Right, the odd part is the mix of 'hand coded streamer and automatic streamer.  I would simply replace
```
Event::Class()->ReadBuffer(buf, this);
```
with
```
...
```

Also note that `TClass::ReadBuffer(TBuffer&, void*)` will read the class version and check the byte count (i.e. fail in this case, since the buffer already advanced passed the version).

Reply:
> Right, the odd part is the mix of 'hand coded streamer and automatic streamer. I would simply replace
> 
> ```
> Event::Class()->ReadBuffer(buf, this);
> ```
> 
> with
> 
> ```
> ...
> ```

I have edited the suggestion and added a more explicit comment.  If there are no other suggestions, we could merge this PR.

Reply:
Go ahead.  It looks good.